nni/algorithms/compression/v2/pytorch/base/compressor.py

```diff
@@ -0,0 +1,282 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import collections
+import logging
+from typing import List, Dict, Optional, OrderedDict, Tuple, Any
+
+import torch
+from torch.nn import Module
+from torch.tensor import Tensor
+
+from nni.common.graph_utils import TorchModuleGraph
+
+_logger = logging.getLogger(__name__)
+
+__all__ = ['LayerInfo', 'Compressor']
+
+
+class LayerInfo:
+    def __init__(self, name: str, module: Module):
+        self.module = module
+        self.name = name
+        self.type = type(module).__name__
+
+
+def _setattr(model: Module, name: str, module: Module):
+    name_list = name.split(".")
+    for name in name_list[:-1]:
+        model = getattr(model, name)
+    setattr(model, name_list[-1], module)
+
+
+weighted_modules = [
+    'Conv1d', 'Conv2d', 'Conv3d', 'ConvTranspose1d', 'ConvTranspose2d', 'ConvTranspose3d',
+    'Linear', 'Bilinear',
+    'PReLU',
+    'Embedding', 'EmbeddingBag',
+]
+
+
+class Compressor:
+    """
+    The abstract base pytorch compressor.
+    """
+
+    def __init__(self, model: Module, config_list: List[Dict]):
+        """
+        Parameters
+        ----------
+        model
+            The model under compressed.
+        config_list
+            The config list used by compressor, usually specifies the 'op_types' or 'op_names' that want to compress.
+        """
+        assert isinstance(model, Module)
+        self.is_wrapped = False
+        self.reset(model=model, config_list=config_list)
+
+    def reset(self, model: Module, config_list: List[Dict]):
+        """
+        Reset the compressor with model and config_list.
+
+        Parameters
+        ----------
+        model
+            The model under compressed.
+        config_list
+            The config list used by compressor, usually specifies the 'op_types' or 'op_names' that want to compress.
+        """
+        assert isinstance(model, Module), 'Only support compressing pytorch Module, but the type of model is {}.'.format(type(model))
+        self.bound_model = model
+        self.config_list = config_list
+        self.validate_config(model=model, config_list=config_list)
+
+        self._unwrap_model()
+
+        self._modules_to_compress = None
+        self.modules_wrapper = collections.OrderedDict()
+        for layer, config in self._detect_modules_to_compress():
+            wrapper = self._wrap_modules(layer, config)
+            self.modules_wrapper[layer.name] = wrapper
+
+        self._wrap_model()
+
+    def _detect_modules_to_compress(self) -> List[Tuple[LayerInfo, Dict]]:
+        """
+        Detect all modules should be compressed, and save the result in `self._modules_to_compress`.
+        The model will be instrumented and user should never edit it after calling this method.
+        """
+        if self._modules_to_compress is None:
```

Comment:
suggest put `self._modules_to_compress=None` into ```__init__```, esle there will be a potential calling order dependency between `_detect_modules_to_compress` and `reset` which may lead to error.

Reply:
I think `self._modules_to_compress=None` in `reset` is enough, because `reset` is a part of logic in `__init__`. And only after set `self.bound_model`, call `_detect_modules_to_compress` is meaningful?

Reply:
Got it.